src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -0,0 +1,3146 @@
+  // output[1]: 299 = 100101011 (with 0 put into place 1, and 1 put into place 4).
+  // output[2]: 313 = 100111001 (with 1 put into place 1, and 0 put into place 4).
+  // output[3]: 313 = 100111011 (with 1's put into places 1 and 4).
+  indexes_t indexes(const reg_t &qubits, const reg_t &qubits_sorted, const uint_t k) const;
+
+  // As above but returns a fixed sized array of of 2^N in ints
+  template<size_t N>
+  areg_t<1ULL << N> indexes(const areg_t<N> &qs, const areg_t<N> &qubits_sorted, const uint_t k) const;
+
+  // State initialization of a component
+  // Initialize the specified qubits to a desired statevector
+  // (leaving the other qubits in their current state)
+  // assuming the qubits being initialized have already been reset to the zero state
+  // (using apply_reset)
+  void initialize_component(const reg_t &qubits, const cvector_t<double> &state);
+
+  //-----------------------------------------------------------------------
+  // Check point operations
+  //-----------------------------------------------------------------------
+
+  // Create a checkpoint of the current state
+  void checkpoint();
+
+  // Revert to the checkpoint
+  void revert(bool keep);
+
+  // Compute the inner product of current state with checkpoint state
+  std::complex<double> inner_product() const;
+
+  //-----------------------------------------------------------------------
+  // Initialization
+  //-----------------------------------------------------------------------
+
+  // Initializes the current vector so that all qubits are in the |0> state.
+  void initialize();
+
+  // Initializes the vector to a custom initial state.
+  // If the length of the data vector does not match the number of qubits
+  // an exception is raised.
+  void initialize_from_vector(const cvector_t<double> &data);
+
+  // Initializes the vector to a custom initial state.
+  // If num_states does not match the number of qubits an exception is raised.
+  void initialize_from_data(const std::complex<data_t>* data, const size_t num_states);
+
+  //-----------------------------------------------------------------------
+  // Apply Matrices
+  //-----------------------------------------------------------------------
+
+  // Apply a 1-qubit matrix to the state vector.
+  // The matrix is input as vector of the column-major vectorized 1-qubit matrix.
+  void apply_matrix(const uint_t qubit, const cvector_t<double> &mat);
+
+  // Apply a N-qubit matrix to the state vector.
+  // The matrix is input as vector of the column-major vectorized N-qubit matrix.
+  void apply_matrix(const reg_t &qubits, const cvector_t<double> &mat);
+
+  // Apply a stacked set of 2^control_count target_count--qubit matrix to the state vector.
+  // The matrix is input as vector of the column-major vectorized N-qubit matrix.
+  void apply_multiplexer(const reg_t &control_qubits, const reg_t &target_qubits, const cvector_t<double> &mat);
```

Comment:
If the gpu doesn't support all the same gates for now as the CPU version, maybe we should add a separate `State` subclass for it that defines a different set of allowed gates. Hopefully this could be done with minimal code copying by something like:

```cpp
namespace AER {
namespace StatevectorThrust {

template <class statevec_t = QV::QubitVectorThrust<double>>
class State : public Statevector::State<statevec_t> {
  // Here we can override gate set methods that are different to
  // CPU statevector class to remove other gates
  
  // Return the string name of the State class
  virtual std::string name() const override {return "statevector_gpu";}

  virtual Operations::OpSet::optypeset_t allowed_ops() const override {
    return // list of osps if different to STatevector state
  }
  virtual stringset_t allowed_gates() const override {
    return // list of gates if different to Statevector::State;
  }

  // other methods should be same and inherited from base class
}
}}
```

Reply:
I added statevector_gpu_state.hpp